-----------

ASSIGNMENT -> SAMPLING UCS654

Name: Jasween Kaur Brar

Roll No: 102017187

Sub-group: 3CS8

-----------

### Import libraries

In [67]:
import pandas as pd
import numpy as np
from collections import Counter
from imblearn.over_sampling import SMOTE

import math

from sklearn.model_selection import train_test_split

from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import GaussianNB
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.ensemble import RandomForestClassifier

from sklearn.metrics import accuracy_score


In [68]:
import warnings
warnings.filterwarnings("ignore")

### Load the dataset

In [69]:
df = pd.read_csv("Creditcard_data.csv")

In [70]:
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0     0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1     0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2     1 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3     1 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4     2 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      1  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

### Split the data into Input Features and Output Target

In [71]:
X = df.iloc[:,:-1]
Y = df.iloc[:,-1]

In [72]:
X.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0     0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1     0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2     1 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3     1 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4     2 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V20       V21       V22       V23       V24  \
0  0.098698  0.363787  ...  0.251412 -0.018307  0.277838 -0.110474  0.066928   
1  0.085102 -0.255425  ... -0.069083 -0.225775 -0.638672  0.101288 -0.339846   
2  0.247676 -1.514654  ...  0.524980  0.247998  0.771679  0.909412 -0.689281   
3  0.377436 -1.387024  ... -0.208038 -0.108300  0.005274 -0.190321 -1.175575   
4 -0.270533  0.817739  ...  0.408542 -0.009431  0.798278 -0.137458  0.141267   

        V25       V26       V27       V28  Amount  
0  0.128539 -0.189115  0.133558 -0.021053  149.62  
1  0.167170  0.125895 -0.008983  0.014724    2.69  
2 -0.327642 -0.139097 -0.055353 -0.059752  378.66  
3  0.647376 -0.221929  0.062723  0.061458  123.50  
4 -0.206010  0.502292  0.219422  0.215153   69.99  

[5 rows x 30 columns]

In [73]:
Y.head()

0    0
1    1
2    0
3    0
4    0
Name: Class, dtype: int64

In [74]:
print("Original dataset shape is %s" % Counter(Y))

Original dataset shape is Counter({0: 763, 1: 9})


* We observe that dataset is highly imbalanced containing only 9 rows with Class:1 and 763 rows with Class:0

### So, let us perform SMOTE oversampling to balance the dataset

In [75]:
sm = SMOTE(random_state=42)
X_new, Y_new = sm.fit_resample(X,Y)

In [76]:
print("Resampled dataset shape is %s" % Counter(Y_new))

Resampled dataset shape is Counter({0: 763, 1: 763})


* Now the dataset is balanced

### SAMPLING TECHNIQUES

* Find out sample size and create sample

In [77]:
data_balanced = pd.concat([X_new, Y_new], axis=1)
data_balanced.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0     0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1     0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2     1 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3     1 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4     2 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      1  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

#### 1) Simple Random Sampling

In [78]:
# Set the confidence level
confidence_level = 0.95

# Set the margin of error
margin_of_error = 0.05

# Calculate the sample size
p = data_balanced['Class'].mean()  # proportion of the target class in the population
z = 1.96  # z-score for 95% confidence level
n = (z**2 * p * (1-p)) / (margin_of_error**2)
n = math.ceil(n)  # round up to the nearest integer
print("Sample size for random sampling:", n)

# Create the simple random sample
sample_simple_random = data_balanced.sample(n, random_state=42)
sample_simple_random.head()

Sample size for random sampling: 385


Time        V1        V2        V3        V4        V5        V6  \
1439   408 -2.263118  1.896864 -1.490959  3.862543 -0.469374 -1.407899   
76      49 -0.549626  0.418949  1.729833  0.203065 -0.187012  0.253878   
1010   381 -2.068375  1.808796 -1.399115  3.601020 -0.384698 -1.255030   
660    499  1.255439  0.307729  0.292700  0.699873 -0.428876 -1.088456   
1132   564  0.540110 -0.265200  0.782693  0.789618  0.263245 -0.701497   

            V7        V8        V9  ...       V21       V22       V23  \
1439 -2.416175  1.336726 -2.676850  ...  0.501235 -0.023823 -0.457858   
76    0.500894  0.251256 -0.227985  ...  0.115062  0.418529 -0.065133   
1010 -2.247747  1.263671 -2.499528  ...  0.451204 -0.070505 -0.402225   
660   0.043840 -0.167739  0.128854  ... -0.294795 -0.882126  0.136846   
1132 -0.391240  0.009643  0.222441  ... -0.130637 -0.376640  0.165142   

           V24       V25       V26       V27       V28     Amount  Class  
1439  0.322511  0.056816  0.155428  0.248174 -0.142655   0.035480      1  
76    0.264981  0.003958  0.395969  0.027182  0.043506  59.990000      0  
1010  0.148055 -0.102013  0.167370  0.258140 -0.105105   0.101191      1  
660   0.327949  0.194459  0.096516 -0.027271  0.029491   1.980000      0  
1132  0.078787  0.124117  0.263234 -0.044749 -0.013546   1.336251      1  

[5 rows x 31 columns]

#### 2) Stratified Sampling

In [79]:
# Set the confidence level
confidence_level = 0.95

# Set the margin of error
margin_of_error = 0.05

# Calculate the sample size
p = data_balanced['Class'].mean()  # proportion of the target class in the population
z = 1.96  # z-score for 95% confidence level
n = (z**2 * p * (1-p)) / (margin_of_error**2)
n = math.ceil(n)  # round up to the nearest integer

strata = data_balanced.groupby('Class')

# sample 2 rows from each stratum
sample_stratified = strata.apply(lambda x: x.sample(n))

sample_stratified
N = len(sample_stratified)
print("Sample size for stratified sampling:", N)
sample_stratified.head()

Sample size for stratified sampling: 770


Time        V1        V2        V3        V4        V5        V6  \
Class                                                                         
0     332   242 -0.398935  0.941953  1.359288  0.013490 -0.191940 -0.824441   
      375   273 -0.792970  0.273688  2.612040  1.213495 -0.225015 -0.594575   
      38     29  0.996370 -0.122589  0.546819  0.706580  0.134560  1.156995   
      171   110 -1.276830 -0.903557  1.477105 -2.582626  0.635981 -0.866537   
      741   557  1.173585  0.250650  0.839631  1.066309 -0.365090 -0.161887   

                 V7        V8        V9  ...       V21       V22       V23  \
Class                                    ...                                 
0     332  0.474948  0.175557 -0.362987  ... -0.222210 -0.704461  0.020807   
      375 -0.250744  0.162628  0.038633  ...  0.136200  0.381140 -0.020689   
      38  -0.294561  0.407429  0.337863  ... -0.076302  0.108867  0.162231   
      171  1.138067 -0.924605 -0.221504  ... -0.482119 -0.182518 -0.614101   
      741 -0.230310 -0.012302  0.123302  ...  0.171834  0.659725 -0.064483   

                V24       V25       V26       V27       V28  Amount  Class  
Class                                                                       
0     332  0.429138 -0.286397  0.048579  0.229096  0.085802    2.69      0  
      375  0.706053 -0.234172 -0.377185  0.164555  0.155042    6.51      0  
      38  -0.575624  0.109795  0.373813  0.050552  0.005106   20.53      0  
      171  0.037650  0.793080 -0.724708 -0.616486 -0.485985  110.40      0  
      741  0.125329  0.506081 -0.270854  0.069604  0.030001    1.00      0  

[5 rows x 31 columns]

#### 3) Cluster Sampling

In [80]:
# Cluster Sampling
def get_clustered_Sample(df, n_per_cluster, num_select_clusters):
    N = len(df)
    K = int(N/n_per_cluster)
    data = None
    for k in range(K):
        sample_k = df.sample(n_per_cluster)
        sample_k["cluster"] = np.repeat(k,len(sample_k))
        df = df.drop(index = sample_k.index)
        data = pd.concat([data,sample_k],axis = 0)
    random_chosen_clusters = np.random.randint(0,K,size = num_select_clusters)
    samples = data[data.cluster.isin(random_chosen_clusters)]
    return(samples)
sample_cluster = get_clustered_Sample(df = data_balanced, n_per_cluster = 20, num_select_clusters = 28)

sample_cluster = sample_cluster.iloc[:,:-1]
N = len(sample_cluster)
print("Sample size for cluster sampling:", N)
sample_cluster.head()


Sample size for cluster sampling: 520


Time        V1        V2        V3        V4        V5        V6  \
945    539 -1.232616 -1.821353  1.958001  1.034130  1.797406  0.202603   
1278   332 -0.446138 -1.136683  1.433565  0.926649  1.109078 -0.200513   
833    547 -0.688384 -1.343610  1.597178  0.959064  1.326414 -0.074958   
1477   200 -0.283361  0.760589  0.143494  0.695777  0.621740  0.000480   
421    305  0.439950 -0.949842 -0.112592  2.572536  0.149829  1.544542   

            V7        V8        V9  ...       V21       V22       V23  \
945  -1.590611  0.552974  0.736233  ...  0.255274  0.720227  0.257728   
1278 -1.061175  0.315076  0.517782  ...  0.083420  0.227027  0.216782   
833  -1.222400  0.386170  0.578497  ...  0.136798  0.383486  0.229304   
1477 -0.127549  0.326796 -0.518996  ... -0.032131 -0.330045  0.058837   
421   0.213718  0.278461 -0.374167  ...  0.136101 -0.275473 -0.554800   

           V24       V25       V26       V27       V28      Amount  Class  
945  -0.574220 -0.115513  0.680331 -0.100479 -0.123449    1.450505      1  
1278 -0.285784 -0.011491  0.495118 -0.076269 -0.074632    2.068202      1  
833  -0.373745 -0.041946  0.552281 -0.083370 -0.089709    1.415428      1  
1477 -1.112049 -1.174641  0.090733  0.236147  0.174311    0.841915      1  
421  -1.278067  0.554427  0.133135 -0.053053  0.058084  379.630000      0  

[5 rows x 31 columns]

#### 4) Systematic Sampling

In [81]:
sample_systematic = data_balanced.iloc[::5]
## Systematically pick every 5th sample

N = len(sample_systematic)
print("Sample size for cluster sampling:", N)
sample_systematic.head()


Sample size for cluster sampling: 306


Time        V1        V2        V3        V4        V5        V6  \
0      0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388   
5      2 -0.425966  0.960523  1.141109 -0.168252  0.420987 -0.029728   
10    10  1.449044 -1.176339  0.913860 -1.375667 -1.971383 -0.629152   
15    12 -0.752417  0.345485  2.057323 -1.468643 -1.158394 -0.077850   
20    16  0.694885 -1.361819  1.029221  0.834159 -1.191209  1.309109   

          V7        V8        V9  ...       V21       V22       V23       V24  \
0   0.239599  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928   
5   0.476201  0.260314 -0.568671  ... -0.208254 -0.559825 -0.026398 -0.371427   
10 -1.423236  0.048456 -1.720408  ... -0.009302  0.313894  0.027740  0.500512   
15 -0.608581  0.003603 -0.436167  ...  0.499625  1.353650 -0.256573 -0.065084   
20 -0.878586  0.445290 -0.446196  ... -0.295583 -0.571955 -0.050881 -0.304215   

         V25       V26       V27       V28  Amount  Class  
0   0.128539 -0.189115  0.133558 -0.021053  149.62      0  
5  -0.232794  0.105915  0.253844  0.081080    3.67      0  
10  0.251367 -0.129478  0.042850  0.016253    7.80      0  
15 -0.039124 -0.087086 -0.180998  0.129394   15.99      0  
20  0.072001 -0.422234  0.086553  0.063499  231.71      0  

[5 rows x 31 columns]

#### 5) Multi-Stage Sampling

First perform cluster sampling and then randomly pick clusters

In [82]:
# Cluster Sampling
def get_clustered_Sample(df, n_per_cluster, num_select_clusters):
    N = len(df)
    K = int(N/n_per_cluster)
    data = None
    for k in range(K):
        sample_k = df.sample(n_per_cluster)
        sample_k["cluster"] = np.repeat(k,len(sample_k))
        df = df.drop(index = sample_k.index)
        data = pd.concat([data,sample_k],axis = 0)
    random_chosen_clusters = np.random.randint(0,K,size = num_select_clusters)
    samples = data[data.cluster.isin(random_chosen_clusters)]
    return(samples)
sample_c = get_clustered_Sample(df = data_balanced, n_per_cluster = 20, num_select_clusters = 28)
sample_c = sample_c.iloc[:,:-1]

confidence_level = 0.95
margin_of_error = 0.05
# Calculate the random sample size
p = data_balanced['Class'].mean()  # proportion of the target class in the population
z = 1.96  # z-score for 95% confidence level
n = (z**2 * p * (1-p)) / (margin_of_error**2)
n = math.ceil(n)  # round up to the nearest integer

# Random sampling
sample_multistage = sample_c.sample(n , random_state=42)
N = len(sample_multistage)
print("Sample size for multi stage sampling:", N)
sample_multistage.head()


Sample size for multi stage sampling: 385


Time        V1        V2        V3        V4        V5        V6  \
1204   538 -1.739924  0.057183  1.180514 -0.645764  0.917248 -0.294443   
787    212 -0.642889  1.148861 -0.763611  2.306812 -0.244826 -0.786179   
715    539 -1.054584  0.671513  2.106718 -0.591701  0.228368 -0.053748   
1395   230 -0.143945  0.978398 -0.447438  1.989092 -0.430574 -1.209975   
1161   487 -2.765590 -2.980932  1.455880  1.982546  1.653541 -0.622679   

            V7        V8        V9  ...       V21       V22       V23  \
1204  0.262518  0.143558  0.016929  ... -0.177915 -0.191669 -0.262355   
787  -1.366123  0.769217 -1.572109  ...  0.163265 -0.322614 -0.195332   
715   0.941606 -0.191062  0.267565  ... -0.049688  0.137518 -0.420259   
1395 -0.942290  0.422447 -1.064926  ...  0.028021 -0.519889 -0.104578   
1161 -0.323551  0.160366  0.056777  ...  0.597938  0.637994  1.088646   

           V24       V25       V26       V27       V28      Amount  Class  
1204 -0.236172 -0.334152  0.240176 -0.343711 -0.383001    0.997661      1  
787   0.005754  0.102950  0.153093  0.132457 -0.068005    1.281510      1  
715   0.130488  0.482469  0.394143 -0.409756 -0.179564   35.640000      0  
1395  0.331886  0.148719  0.127306  0.088212 -0.037408    1.635113      1  
1161 -0.443922  0.146783  0.122837 -0.218621 -0.019354  388.422097      1  

[5 rows x 31 columns]

#### Apply 5 different models on all 5 samples
* DecisionTreeClassifier
* LogisticRegression
* GaussianNB
* GradientBoostingClassifier
* RandomForestClassifier

In [83]:
# define the samples
samples = {"Sampling 1": sample_simple_random,
            "Sampling 2": sample_stratified,
            "Sampling 3": sample_cluster,
            "Sampling 4": sample_systematic,
            "Sampling 5": sample_multistage}


models = {'M1 = Decision Tree': DecisionTreeClassifier(random_state=0),
    'M2 = Logistic Regression': LogisticRegression(random_state=0),
    'M3 = Gaussian Naive Bayes': GaussianNB(),
    'M4 = Gradient Boosting Classifier': GradientBoostingClassifier(random_state=0),
    'M5 = Random Forest': RandomForestClassifier(random_state=0)}


In [85]:
results=pd.DataFrame(columns= [sampling_name for sampling_name, _ in samples.items()], index= [model_name for model_name, _ in models.items()])

for sampling_name, sample in samples.items():
    for model_name, model in models.items():
        
        X_sample = sample.iloc[:,:-1]
        Y_sample = sample.iloc[:,-1]

        # Split the sample into training and testing sets
        X_train, X_test, y_train, y_test = train_test_split(X_sample, Y_sample, test_size=0.2, random_state=42)

        # Train machine learning model on the training set
        model.fit(X_train, y_train)

        # Make predictions on the testing set
        y_pred = model.predict(X_test)

        # Calculate the accuracy score of the model
        accuracy = accuracy_score(y_test, y_pred)
        results.loc[model_name,sampling_name] = accuracy

In [86]:
results

Sampling 1 Sampling 2 Sampling 3 Sampling 4  \
M1 = Decision Tree                  0.935065   0.967532   0.913462   0.887097   
M2 = Logistic Regression            0.896104   0.915584   0.913462   0.741935   
M3 = Gaussian Naive Bayes           0.883117   0.850649   0.894231   0.790323   
M4 = Gradient Boosting Classifier   0.961039   0.987013   0.951923   0.935484   
M5 = Random Forest                  0.974026   0.993506   0.980769   0.983871   

                                  Sampling 5  
M1 = Decision Tree                  0.961039  
M2 = Logistic Regression            0.974026  
M3 = Gaussian Naive Bayes           0.818182  
M4 = Gradient Boosting Classifier   0.974026  
M5 = Random Forest                  0.987013